File: sdk/mindmodel-client.js
/**
 * MindModel JavaScript/Node.js SDK Client
 * 
 * A JavaScript client library for interacting with the MindModel API.
 */

class MindModelClient {
    /**
     * Create a new MindModel client.
     * 
     * @param {string} baseUrl - The base URL of the MindModel API
     * @param {number} timeout - Request timeout in milliseconds
     */
    constructor(baseUrl = 'http://10.11.2.6:8000', timeout = 30000) {
        this.baseUrl = baseUrl.replace(/\/$/, '');
        this.timeout = timeout;
    }

    /**
     * Make an HTTP request to the API.
     * 
     * @param {string} method - HTTP method
     * @param {string} endpoint - API endpoint path
     * @param {Object} options - Request options
     * @returns {Promise<Object>} The response data
     * @throws {MindModelError} If the request fails
     */
    async _makeRequest(method, endpoint, options = {}) {
        const url = `${this.baseUrl}${endpoint}`;
        const config = {
            method,
            headers: {
                'Content-Type': 'application/json',
                ...options.headers
            },
            timeout: this.timeout,
            ...options
        };

        try {
            const response = await fetch(url, config);
            
            if (!response.ok) {
                const errorText = await response.text();
                throw new MindModelError(`HTTP ${response.status}: ${errorText}`);
            }
            
            return await response.json();
        } catch (error) {
            if (error instanceof MindModelError) {
                throw error;
            }
            throw new MindModelError(`API request failed: ${error.message}`);
        }
    }

    /**
     * Generate a conclusion from input text.
     * 
     * @param {string} inputText - The text to generate a conclusion from
     * @param {number} targetLength - Target length for the conclusion (8-50)
     * @param {string} lengthTag - Optional length tag for the conclusion
     * @returns {Promise<Object>} Object containing conclusion, length, and confidence
     * @throws {MindModelError} If the request fails
     * @throws {Error} If input parameters are invalid
     */
    async generateConclusion(inputText, targetLength = 24, lengthTag = null) {
        if (!inputText || !inputText.trim()) {
            throw new Error('Input text cannot be empty');
        }

        if (targetLength < 8 || targetLength > 50) {
            throw new Error('Target length must be between 8 and 50');
        }

        const payload = {
            input: inputText,
            target_length: targetLength
        };

        if (lengthTag) {
            payload.length_tag = lengthTag;
        }

        return await this._makeRequest('POST', '/v1/conclude', {
            body: JSON.stringify(payload)
        });
    }

    /**
     * Train the model with new examples.
     * 
     * @param {Array<Object>} examples - List of training examples
     * @param {number} epochs - Number of training epochs (1-10)
     * @param {number} batchSize - Training batch size (1-32)
     * @param {string} modelName - Optional base model name
     * @returns {Promise<Object>} Object containing training ID, status, and message
     * @throws {MindModelError} If the request fails
     * @throws {Error} If input parameters are invalid
     */
    async trainModel(examples, epochs = 1, batchSize = 8, modelName = null) {
        if (!examples || examples.length === 0) {
            throw new Error('At least one training example is required');
        }

        if (epochs < 1 || epochs > 10) {
            throw new Error('Epochs must be between 1 and 10');
        }

        if (batchSize < 1 || batchSize > 32) {
            throw new Error('Batch size must be between 1 and 32');
        }

        // Validate examples
        for (let i = 0; i < examples.length; i++) {
            const example = examples[i];
            if (typeof example !== 'object' || example === null) {
                throw new Error(`Example ${i} must be an object`);
            }

            if (!example.input || !example.target) {
                throw new Error(`Example ${i} must contain 'input' and 'target' fields`);
            }

            if (!example.input.trim() || !example.target.trim()) {
                throw new Error(`Example ${i} input and target cannot be empty`);
            }
        }

        const payload = {
            examples,
            epochs,
            batch_size: batchSize
        };

        if (modelName) {
            payload.model_name = modelName;
        }

        return await this._makeRequest('POST', '/v1/learn', {
            body: JSON.stringify(payload)
        });
    }

    /**
     * Get the status of a training job.
     * 
     * @param {string} trainingId - The ID of the training job
     * @returns {Promise<Object>} Object containing training status information
     * @throws {MindModelError} If the request fails
     */
    async getTrainingStatus(trainingId) {
        return await this._makeRequest('GET', `/v1/training/${trainingId}`);
    }

    /**
     * Wait for a training job to complete.
     * 
     * @param {string} trainingId - The ID of the training job
     * @param {number} pollInterval - How often to check status (milliseconds)
     * @param {number} timeout - Maximum time to wait (milliseconds, null for no timeout)
     * @returns {Promise<Object>} Final training status
     * @throws {MindModelError} If the request fails or timeout occurs
     */
    async waitForTrainingCompletion(trainingId, pollInterval = 5000, timeout = null) {
        const startTime = Date.now();

        while (true) {
            const status = await this.getTrainingStatus(trainingId);

            if (status.status === 'completed' || status.status === 'failed') {
                return status;
            }

            if (timeout && (Date.now() - startTime) > timeout) {
                throw new MindModelError(`Training timeout after ${timeout} milliseconds`);
            }

            await new Promise(resolve => setTimeout(resolve, pollInterval));
        }
    }

    /**
     * Get a list of available trained models.
     * 
     * @returns {Promise<Object>} Object containing list of models
     * @throws {MindModelError} If the request fails
     */
    async listModels() {
        return await this._makeRequest('GET', '/v1/models');
    }

    /**
     * Load a specific trained model.
     * 
     * @param {string} modelId - The ID of the model to load
     * @returns {Promise<Object>} Object containing load status information
     * @throws {MindModelError} If the request fails
     */
    async loadModel(modelId) {
        return await this._makeRequest('POST', `/v1/models/${modelId}/load`);
    }

    /**
     * Get the current system status.
     * 
     * @returns {Promise<Object>} Object containing system status information
     * @throws {MindModelError} If the request fails
     */
    async getStatus() {
        return await this._makeRequest('GET', '/status');
    }

    /**
     * Check if the service is healthy.
     * 
     * @returns {Promise<Object>} Object containing health status
     * @throws {MindModelError} If the service is unhealthy
     */
    async healthCheck() {
        return await this._makeRequest('GET', '/health');
    }

    /**
     * Check if the service is healthy (returns boolean).
     * 
     * @returns {Promise<boolean>} True if healthy, False otherwise
     */
    async isHealthy() {
        try {
            await this.healthCheck();
            return true;
        } catch (error) {
            return false;
        }
    }
}

/**
 * Custom error class for MindModel API errors.
 */
class MindModelError extends Error {
    constructor(message) {
        super(message);
        this.name = 'MindModelError';
    }
}

/**
 * Quick function to generate a conclusion from text.
 * 
 * @param {string} inputText - The text to generate a conclusion from
 * @param {string} baseUrl - The base URL of the MindModel API
 * @returns {Promise<string>} The generated conclusion
 * @throws {MindModelError} If the request fails
 */
async function quickConclude(inputText, baseUrl = 'http://10.11.2.6:8000') {
    const client = new MindModelClient(baseUrl);
    const result = await client.generateConclusion(inputText);
    return result.conclusion;
}

/**
 * Train the model with examples and optionally wait for completion.
 * 
 * @param {Array<Object>} examples - List of training examples
 * @param {string} baseUrl - The base URL of the MindModel API
 * @param {boolean} waitForCompletion - Whether to wait for training to complete
 * @returns {Promise<Object>} Training status information
 * @throws {MindModelError} If the request fails
 */
async function batchTrain(examples, baseUrl = 'http://10.11.2.6:8000', waitForCompletion = true) {
    const client = new MindModelClient(baseUrl);
    const result = await client.trainModel(examples);
    
    if (waitForCompletion) {
        return await client.waitForTrainingCompletion(result.training_id);
    }
    
    return result;
}

// Export for different module systems
if (typeof module !== 'undefined' && module.exports) {
    // Node.js
    module.exports = {
        MindModelClient,
        MindModelError,
        quickConclude,
        batchTrain
    };
} else if (typeof window !== 'undefined') {
    // Browser
    window.MindModelClient = MindModelClient;
    window.MindModelError = MindModelError;
    window.quickConclude = quickConclude;
    window.batchTrain = batchTrain;
}

// Example usage (when run directly)
if (typeof require !== 'undefined' && require.main === module) {
    async function example() {
        const client = new MindModelClient();
        
        try {
            // Generate a conclusion
            const result = await client.generateConclusion(
                "The company reported $3.2M revenue in Q2 2024, representing a 15% increase over Q1."
            );
            console.log(`Conclusion: ${result.conclusion}`);
            console.log(`Length: ${result.length}`);
            console.log(`Confidence: ${(result.confidence * 100).toFixed(1)}%`);
            
            // Train the model
            const examples = [
                {
                    input: "The company reported $3.2M revenue in Q2 2024, representing a 15% increase over Q1.",
                    target: "Company revenue increased 15% to $3.2M in Q2.",
                    length_tag: "<LEN_24>",
                    phenomena_tags: ["numbers", "percentage"]
                }
            ];
            
            const trainingResult = await client.trainModel(examples);
            console.log(`Training started: ${trainingResult.training_id}`);
            
            // Wait for completion
            const finalStatus = await client.waitForTrainingCompletion(trainingResult.training_id);
            console.log(`Training completed: ${finalStatus.status}`);
            
            // Check system status
            const status = await client.getStatus();
            console.log(`Model loaded: ${status.model_loaded}`);
            console.log(`Currently training: ${status.is_training}`);
            
        } catch (error) {
            console.error(`Error: ${error.message}`);
        }
    }
    
    example();
}

File: sdk/mindmodel_client.py
"""
MindModel Python SDK Client

A Python client library for interacting with the MindModel API.
"""

import requests
import json
from typing import List, Dict, Optional, Union
from datetime import datetime
import time


class MindModelClient:
    """
    Client for interacting with the MindModel API.
    
    This client provides methods for generating conclusions, training the model,
    and managing the system status.
    """
    
    def __init__(self, base_url: str = "http://10.11.2.6:8000", timeout: int = 30):
        """
        Initialize the MindModel client.
        
        Args:
            base_url: The base URL of the MindModel API
            timeout: Request timeout in seconds
        """
        self.base_url = base_url.rstrip('/')
        self.timeout = timeout
        self.session = requests.Session()
        
    def _make_request(self, method: str, endpoint: str, **kwargs) -> requests.Response:
        """
        Make an HTTP request to the API.
        
        Args:
            method: HTTP method (GET, POST, etc.)
            endpoint: API endpoint path
            **kwargs: Additional arguments for requests
            
        Returns:
            requests.Response: The HTTP response
            
        Raises:
            requests.RequestException: If the request fails
        """
        url = f"{self.base_url}{endpoint}"
        kwargs.setdefault('timeout', self.timeout)
        
        try:
            response = self.session.request(method, url, **kwargs)
            response.raise_for_status()
            return response
        except requests.RequestException as e:
            raise MindModelError(f"API request failed: {e}")
    
    def generate_conclusion(
        self, 
        input_text: str, 
        target_length: int = 24, 
        length_tag: Optional[str] = None
    ) -> Dict[str, Union[str, int, float]]:
        """
        Generate a conclusion from input text.
        
        Args:
            input_text: The text to generate a conclusion from
            target_length: Target length for the conclusion (8-50)
            length_tag: Optional length tag for the conclusion
            
        Returns:
            Dict containing the conclusion, length, and confidence
            
        Raises:
            MindModelError: If the request fails
            ValueError: If input parameters are invalid
        """
        if not input_text.strip():
            raise ValueError("Input text cannot be empty")
        
        if not (8 <= target_length <= 50):
            raise ValueError("Target length must be between 8 and 50")
        
        payload = {
            "input": input_text,
            "target_length": target_length
        }
        
        if length_tag:
            payload["length_tag"] = length_tag
        
        response = self._make_request(
            "POST", 
            "/v1/conclude", 
            json=payload
        )
        
        return response.json()
    
    def train_model(
        self, 
        examples: List[Dict], 
        epochs: int = 1, 
        batch_size: int = 8,
        model_name: Optional[str] = None
    ) -> Dict[str, str]:
        """
        Train the model with new examples.
        
        Args:
            examples: List of training examples
            epochs: Number of training epochs (1-10)
            batch_size: Training batch size (1-32)
            model_name: Optional base model name
            
        Returns:
            Dict containing training ID, status, and message
            
        Raises:
            MindModelError: If the request fails
            ValueError: If input parameters are invalid
        """
        if not examples:
            raise ValueError("At least one training example is required")
        
        if not (1 <= epochs <= 10):
            raise ValueError("Epochs must be between 1 and 10")
        
        if not (1 <= batch_size <= 32):
            raise ValueError("Batch size must be between 1 and 32")
        
        # Validate examples
        for i, example in enumerate(examples):
            if not isinstance(example, dict):
                raise ValueError(f"Example {i} must be a dictionary")
            
            if 'input' not in example or 'target' not in example:
                raise ValueError(f"Example {i} must contain 'input' and 'target' fields")
            
            if not example['input'].strip() or not example['target'].strip():
                raise ValueError(f"Example {i} input and target cannot be empty")
        
        payload = {
            "examples": examples,
            "epochs": epochs,
            "batch_size": batch_size
        }
        
        if model_name:
            payload["model_name"] = model_name
        
        response = self._make_request(
            "POST", 
            "/v1/learn", 
            json=payload
        )
        
        return response.json()
    
    def get_training_status(self, training_id: str) -> Dict[str, Union[str, int, datetime]]:
        """
        Get the status of a training job.
        
        Args:
            training_id: The ID of the training job
            
        Returns:
            Dict containing training status information
            
        Raises:
            MindModelError: If the request fails
        """
        response = self._make_request("GET", f"/v1/training/{training_id}")
        return response.json()
    
    def wait_for_training_completion(
        self, 
        training_id: str, 
        poll_interval: int = 5, 
        timeout: Optional[int] = None
    ) -> Dict[str, Union[str, int, datetime]]:
        """
        Wait for a training job to complete.
        
        Args:
            training_id: The ID of the training job
            poll_interval: How often to check status (seconds)
            timeout: Maximum time to wait (seconds, None for no timeout)
            
        Returns:
            Final training status
            
        Raises:
            MindModelError: If the request fails or timeout occurs
        """
        start_time = time.time()
        
        while True:
            status = self.get_training_status(training_id)
            
            if status['status'] in ['completed', 'failed']:
                return status
            
            if timeout and (time.time() - start_time) > timeout:
                raise MindModelError(f"Training timeout after {timeout} seconds")
            
            time.sleep(poll_interval)
    
    def list_models(self) -> Dict[str, List[Dict]]:
        """
        Get a list of available trained models.
        
        Returns:
            Dict containing list of models
            
        Raises:
            MindModelError: If the request fails
        """
        response = self._make_request("GET", "/v1/models")
        return response.json()
    
    def load_model(self, model_id: str) -> Dict[str, str]:
        """
        Load a specific trained model.
        
        Args:
            model_id: The ID of the model to load
            
        Returns:
            Dict containing load status information
            
        Raises:
            MindModelError: If the request fails
        """
        response = self._make_request("POST", f"/v1/models/{model_id}/load")
        return response.json()
    
    def get_status(self) -> Dict[str, Union[bool, int, str]]:
        """
        Get the current system status.
        
        Returns:
            Dict containing system status information
            
        Raises:
            MindModelError: If the request fails
        """
        response = self._make_request("GET", "/status")
        return response.json()
    
    def health_check(self) -> Dict[str, str]:
        """
        Check if the service is healthy.
        
        Returns:
            Dict containing health status
            
        Raises:
            MindModelError: If the service is unhealthy
        """
        response = self._make_request("GET", "/health")
        return response.json()
    
    def is_healthy(self) -> bool:
        """
        Check if the service is healthy (returns boolean).
        
        Returns:
            True if healthy, False otherwise
        """
        try:
            self.health_check()
            return True
        except MindModelError:
            return False
    
    def close(self):
        """Close the client session."""
        self.session.close()
    
    def __enter__(self):
        return self
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        self.close()


class MindModelError(Exception):
    """Exception raised for MindModel API errors."""
    pass


# Convenience functions for common operations

def quick_conclude(input_text: str, base_url: str = "http://10.11.2.6:8000") -> str:
    """
    Quick function to generate a conclusion from text.
    
    Args:
        input_text: The text to generate a conclusion from
        base_url: The base URL of the MindModel API
        
    Returns:
        The generated conclusion
        
    Raises:
        MindModelError: If the request fails
    """
    with MindModelClient(base_url) as client:
        result = client.generate_conclusion(input_text)
        return result['conclusion']


def batch_train(
    examples: List[Dict], 
    base_url: str = "http://10.11.2.6:8000",
    wait_for_completion: bool = True
) -> Dict[str, Union[str, int, datetime]]:
    """
    Train the model with examples and optionally wait for completion.
    
    Args:
        examples: List of training examples
        base_url: The base URL of the MindModel API
        wait_for_completion: Whether to wait for training to complete
        
    Returns:
        Training status information
        
    Raises:
        MindModelError: If the request fails
    """
    with MindModelClient(base_url) as client:
        result = client.train_model(examples)
        
        if wait_for_completion:
            return client.wait_for_training_completion(result['training_id'])
        
        return result


# Example usage and documentation

if __name__ == "__main__":
    # Example usage
    client = MindModelClient()
    
    try:
        # Generate a conclusion
        result = client.generate_conclusion(
            "The company reported $3.2M revenue in Q2 2024, representing a 15% increase over Q1."
        )
        print(f"Conclusion: {result['conclusion']}")
        print(f"Length: {result['length']}")
        print(f"Confidence: {result['confidence']:.2%}")
        
        # Train the model
        examples = [
            {
                "input": "The company reported $3.2M revenue in Q2 2024, representing a 15% increase over Q1.",
                "target": "Company revenue increased 15% to $3.2M in Q2.",
                "length_tag": "<LEN_24>",
                "phenomena_tags": ["numbers", "percentage"]
            }
        ]
        
        training_result = client.train_model(examples)
        print(f"Training started: {training_result['training_id']}")
        
        # Wait for completion
        final_status = client.wait_for_training_completion(training_result['training_id'])
        print(f"Training completed: {final_status['status']}")
        
        # Check system status
        status = client.get_status()
        print(f"Model loaded: {status['model_loaded']}")
        print(f"Currently training: {status['is_training']}")
        
    except MindModelError as e:
        print(f"Error: {e}")
    finally:
        client.close()
